src/wallet/db.cpp

```diff
@@ -3,23 +3,125 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <chainparams.h>
 #include <fs.h>
+#include <logging.h>
 #include <wallet/db.h>
 
 #include <string>
 
-void SplitWalletPath(const fs::path& wallet_path, fs::path& env_directory, std::string& database_filename)
+std::vector<fs::path> ListDatabases(const fs::path& wallet_dir)
 {
-    if (fs::is_regular_file(wallet_path)) {
+    const size_t offset = wallet_dir.string().size() + 1;
+    std::vector<fs::path> paths;
+    boost::system::error_code ec;
+
+    for (auto it = fs::recursive_directory_iterator(wallet_dir, ec); it != fs::recursive_directory_iterator(); it.increment(ec)) {
+        if (ec) {
+            LogPrintf("%s: %s %s\n", __func__, ec.message(), it->path().string());
+            continue;
+        }
+
+        // Get wallet path relative to walletdir by removing walletdir from the wallet path.
+        // This can be replaced by boost::filesystem::lexically_relative once boost is bumped to 1.60.
+        const fs::path path = it->path().string().substr(offset);
+
+        if (it->status().type() == fs::directory_file &&
+            (IsBDBFile(BDBDataFile(it->path())) || IsSQLiteFile(SQLiteDataFile(it->path())))) {
+            // Found a directory which contains wallet.dat btree file, add it as a wallet.
+            paths.emplace_back(path);
+        } else if (it.level() == 0 && it->symlink_status().type() == fs::regular_file && IsBDBFile(it->path())) {
+            if (it->path().filename() == "wallet.dat") {
+                // Found top-level wallet.dat btree file, add top level directory ""
+                // as a wallet.
+                paths.emplace_back();
+            } else {
+                // Found top-level btree file not called wallet.dat. Current bitcoin
+                // software will never create these files but will allow them to be
+                // opened in a shared database environment for backwards compatibility.
+                // Add it to the list of available wallets.
+                paths.emplace_back(path);
+            }
+        }
+    }
+
+    return paths;
+}
+
+fs::path BDBDataFile(const fs::path& path)
+{
+    if (fs::is_regular_file(path)) {
         // Special case for backwards compatibility: if wallet path points to an
         // existing file, treat it as the path to a BDB data file in a parent
         // directory that also contains BDB log files.
-        env_directory = wallet_path.parent_path();
-        database_filename = wallet_path.filename().string();
+        return path;
     } else {
         // Normal case: Interpret wallet path as a directory path containing
         // data and log files.
-        env_directory = wallet_path;
-        database_filename = "wallet.dat";
+        return path / "wallet.dat";
     }
 }
+
+fs::path SQLiteDataFile(const fs::path& path)
+{
+    return path / "wallet.dat";
+}
+
+bool IsBDBFile(const fs::path& path)
+{
+    if (!fs::exists(path)) return false;
```

Comment:
Is this check necessary?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/20275#discussion_r515673173

> Is this check necessary?

This is preexisting code in a move-only function. If you wanted to drop this line you would need to add an error_code check below to avoid logging an error if the path is not found. That's probably the way I would do it, but I don't think the difference is significant.